node/src/reactor/main_reactor/control.rs

```diff
@@ -736,135 +722,181 @@ impl MainReactor {
+                        }
+                        LeapStatus::Awaiting { .. } => KeepUpInstruction::CheckLater(
+                            "Historical: sync leap awaiting".to_string(),
+                            self.control_logic_default_delay.into(),
+                        ),
                     }
-                }
-                BlockSynchronizerProgress::Syncing(_, maybe_block_height, last_progress) => {
-                    error!(
-                        "XXX attempting historical syncing maybe_block_height: {:?}",
-                        maybe_block_height
+                } else if self.block_synchronizer.register_block_by_hash(
+                    parent_hash,
+                    true,
+                    true,
+                    self.chainspec.core_config.simultaneous_peer_requests,
+                ) {
+                    debug!("Historical: register_block_by_hash: {:?}", parent_hash);
+                    let peers_to_ask = self.small_network.fully_connected_peers_random(
+                        rng,
+                        self.chainspec.core_config.simultaneous_peer_requests as usize,
                     );
-                    if self.last_progress < last_progress {
-                        self.last_progress = last_progress;
-                    }
+                    debug!("Historical: peers count: {:?}", peers_to_ask.len());
+                    self.block_synchronizer
+                        .register_peers(parent_hash, peers_to_ask);
+
+                    return KeepUpInstruction::Do(
+                        Duration::ZERO,
+                        effect_builder.immediately().event(|_| {
+                            MainEvent::BlockSynchronizerRequest(BlockSynchronizerRequest::NeedNext)
+                        }),
+                    );
+                } else {
+                    KeepUpInstruction::CheckLater(
+                        format!("Historical: has {:?} but is stuck for some reason", era_id),
+                        self.control_logic_default_delay.into(),
+                    )
                 }
             }
         }
-        KeepUpInstruction::CheckLater(
-            "KeepUp: at perceived tip of chain".to_string(),
-            self.control_logic_default_delay.into(),
-        )
+    }
+
+    fn update_last_progress(
+        &mut self,
+        block_synchronizer_progress: &BlockSynchronizerProgress,
+        phase_prefix: &str,
+    ) {
+        if let BlockSynchronizerProgress::Syncing(_, _, last_progress) = block_synchronizer_progress
+        {
+            // do idleness / reattempt checking
+            let sync_progress = *last_progress;
+            if sync_progress > self.last_progress {
+                self.last_progress = sync_progress;
+                // if any progress has been made, reset attempts
+                self.attempts = 0;
+            }
+            if Timestamp::now().saturating_diff(self.last_progress) > self.idle_tolerance {
```

Comment:
I think we have an `elapsed` method:
```suggestion
            if self.last_progress.elapsed() > self.idle_tolerance {
```

Reply:
applied